src/libexpr/eval.cc

```diff
@@ -716,41 +716,51 @@ Value * ExprPath::maybeThunk(EvalState & state, Env & env)
 
 void EvalState::evalFile(const Path & path_, Value & v)
 {
-    auto path = checkSourcePath(path_);
+    if (path_ == "-") {
+        Expr * e = parseStdin();
+        try {
+            eval(e, v);
+        } catch (Error & e) {
+            addErrorPrefix(e, "while evaluating %1%:\n", "<stdin>");
+            throw;
+        }
+    } else {
+        auto path = checkSourcePath(path_);
 
-    FileEvalCache::iterator i;
```

Comment:
Pre-existing problem, but while looking through this... declaring these temporaries in the conditional seems to solve obvious scoping problems currently resolved by using variable names like `path2` for temporaries later in the sequence... depending on `auto` preferences it'd look like this:

(EDIT: just kidding, looks like a clean solution for this requires c++17 maybe? (`if (auto i = fileEvalCache.find(path); i != fileEvalCache.end())`))